<strong>Importación de librerias</strong> 

In [555]:
import pandas as pd
import sqlite3 as sql3

<strong>Trasnformamos los csv de usuarios_ecobici-#anios.csv en df para depues tener un buen analisis de datos </strong>

In [556]:
userBici15 = pd.read_csv("usuarios-ecobici-2015.csv")
userBici16 = pd.read_csv("usuarios-ecobici-2016.csv")
userBici17 = pd.read_csv("usuarios-ecobici-2017.csv")
userBici18 = pd.read_csv("usuarios-ecobici-2018.csv")
userBici19 = pd.read_csv("usuarios_ecobici_2019.csv")
userBici20 = pd.read_csv("usuarios_ecobici_2020.csv")
userBici21 = pd.read_csv("usuarios_ecobici_2021.csv")
userBici22 = pd.read_csv("usuarios_ecobici_2022.csv")
userBici23 = pd.read_csv("usuarios_ecobici_2023.csv",low_memory=False) # SE AGREGA EL PARAMETRO LOW_MEMORY PARA EVITAR CONFLICTOS CON DATOS MIXTOS
userBici24 = pd.read_csv("usuarios_ecobici_2024.csv")

<strong>Revisamos cuantos registros vienen en cada df</strong>

In [557]:
print(f"El archivo 2015 tiene {userBici15.shape[0]} filas")
print(f"El archivo 2016 tiene {userBici16.shape[0]} filas")
print(f"El archivo 2017 tiene {userBici17.shape[0]} filas")
print(f"El archivo 2018 tiene {userBici18.shape[0]} filas") 
print(f"El archivo 2019 tiene {userBici19.shape[0]} filas")
print (f"El archivo 2020 tiene {userBici20.shape[0]} filas")
print (f"El archivo 2021 tiene {userBici21.shape[0]} filas")
print (f"El archivo 2022 tiene {userBici22.shape[0]} filas")
print (f"El archivo 2023 tiene {userBici23.shape[0]} filas")
print (f"El archivo 2024 tiene {userBici24.shape[0]} filas")


El archivo 2015 tiene 40397 filas
El archivo 2016 tiene 39481 filas
El archivo 2017 tiene 58222 filas
El archivo 2018 tiene 56182 filas
El archivo 2019 tiene 83600 filas
El archivo 2020 tiene 107166 filas
El archivo 2021 tiene 81577 filas
El archivo 2022 tiene 105997 filas
El archivo 2023 tiene 136066 filas
El archivo 2024 tiene 89504 filas


<strong>Reviso el tipo de dato de todas la columnas de los df<strong>

In [558]:
#Esta función tiene dos input:
#1.Es una lista de los df  y revisando uno por uno las columnas ; 
#2.Deja los nombres de los df
# Me retorna un df donde me muestra el tipo de dato de c/columna 
def leer_formatos_columnas(lista_df, nombres_df):
    resultados = []
    for nombre, df in zip(nombres_df, lista_df):
        for columna, tipo in df.dtypes.items():
            resultados.append({'DataFrame': nombre, 'Columna': columna, 'Tipo': tipo})
    df_resultados = pd.DataFrame(resultados)
    return df_resultados

listDfUserBici = [userBici15,userBici16, userBici17, userBici18, userBici19, userBici20, userBici21, userBici22, userBici23, userBici24]
listNameUserBici = ['userBici15','userBici16','userBici17','userBici18','userBici19','userBici20', 'userBici21', 'userBici22', 'userBici23', 'userBici24']

leer_formatos_columnas(listDfUserBici, listNameUserBici)


DataFrame                      Columna     Tipo
0   userBici15                   id_usuario    int64
1   userBici15               genero_usuario   object
2   userBici15                 edad_usuario    int64
3   userBici15                   fecha_alta   object
4   userBici15                    hora_alta   object
5   userBici16                   id_usuario    int64
6   userBici16               genero_usuario   object
7   userBici16                 edad_usuario    int64
8   userBici16                   fecha_alta   object
9   userBici16                    hora_alta   object
10  userBici17                   id_usuario    int64
11  userBici17               genero_usuario   object
12  userBici17                 edad_usuario    int64
13  userBici17                   fecha_alta   object
14  userBici17                    hora_alta   object
15  userBici18                   id_usuario    int64
16  userBici18               genero_usuario   object
17  userBici18                 edad_usuario    int64
18  userBici18                   fecha_alta   object
19  userBici18                    hora_alta   object
20  userBici19                   ID_usuario    int64
21  userBici19               genero_usuario   object
22  userBici19                 edad_usuario  float64
23  userBici19                   fecha_alta   object
24  userBici19                    hora_alta   object
25  userBici20                   ID_usuario    int64
26  userBici20               genero_usuario   object
27  userBici20                 edad_usuario  float64
28  userBici20                   fecha_alta   object
29  userBici20                    hora_alta   object
30  userBici20  Customer.Has.Dni..Yes...No.   object
31  userBici21                   ID_usuario    int64
32  userBici21               genero_usuario   object
33  userBici21                 edad_usuario   object
34  userBici21                   fecha_alta   object
35  userBici21                    hora_alta   object
36  userBici21  Customer.Has.Dni..Yes...No.   object
37  userBici22                   ID_usuario    int64
38  userBici22               genero_usuario   object
39  userBici22                 edad_usuario   object
40  userBici22                   fecha_alta   object
41  userBici22                    hora_alta   object
42  userBici22  Customer.Has.Dni..Yes...No.   object
43  userBici23                   ID_usuario    int64
44  userBici23               genero_usuario   object
45  userBici23                 edad_usuario   object
46  userBici23                   fecha_alta   object
47  userBici23                    hora_alta   object
48  userBici23  Customer.Has.Dni..Yes...No.   object
49  userBici24                   ID_usuario    int64
50  userBici24               genero_usuario   object
51  userBici24                 edad_usuario   object
52  userBici24                   fecha_alta   object
53  userBici24                    hora_alta   object
54  userBici24  Customer.Has.Dni..Yes...No.   object

<strong>En userBici19, userBici20, userBici21, userBici22, userBici23, userBici24 en la columna 'edad_usuario' esta en float lo cambiamos a int <strong>

In [559]:
#Definimos la función replace_format
def replace_format (list_df, column):
    for df in list_df:
        df[column] = pd.to_numeric(df[column], errors='coerce')
        df[column] = df[column].fillna(0).astype('int64')
    return 
edad_usuario = 'edad_usuario'
replace_format([userBici19, userBici20, userBici21, userBici22, userBici23, userBici24], edad_usuario)
#Verificamos en un df 
userBici19.dtypes

ID_usuario         int64
genero_usuario    object
edad_usuario       int64
fecha_alta        object
hora_alta         object
dtype: object

<strong>Revisamos las columna edad_usuario en los df <strong>

In [560]:
#Esta funcion hace un filtrado a los df para determinar la 17 < edad > 85
def analizar_edad_usuario(df):
    df['edad_usuario'] = pd.to_numeric(df['edad_usuario'], errors='coerce')
    # Filtrar los registros con edad_usuario menor a 17 o mayor a 85
    df_filtrado = df[(df['edad_usuario'] < 17) | (df['edad_usuario'] > 85)]
    cantidad_filtrados = df_filtrado.shape[0]    
    return cantidad_filtrados

# Aplica la funcion y muestra los resultados
print(f"Cantidad de registros filtrados en userBici15: {analizar_edad_usuario(userBici15)}")
print(f"Cantidad de registros filtrados en userBici16: {analizar_edad_usuario(userBici16)}")
print(f"Cantidad de registros filtrados en userBici17: {analizar_edad_usuario(userBici17)}")
print(f"Cantidad de registros filtrados en userBici18: {analizar_edad_usuario(userBici18)}")
print(f"Cantidad de registros filtrados en userBici19: {analizar_edad_usuario(userBici19)}")
print(f"Cantidad de registros filtrados en userBici20: {analizar_edad_usuario(userBici20)}")
print(f"Cantidad de registros filtrados en userBici21: {analizar_edad_usuario(userBici21)}")
print(f"Cantidad de registros filtrados en userBici22: {analizar_edad_usuario(userBici22)}")
print(f"Cantidad de registros filtrados en userBici23: {analizar_edad_usuario(userBici23)}")
print(f"Cantidad de registros filtrados en userBici24: {analizar_edad_usuario(userBici24)}")

Cantidad de registros filtrados en userBici15: 22
Cantidad de registros filtrados en userBici16: 11
Cantidad de registros filtrados en userBici17: 120
Cantidad de registros filtrados en userBici18: 243
Cantidad de registros filtrados en userBici19: 822
Cantidad de registros filtrados en userBici20: 85
Cantidad de registros filtrados en userBici21: 32
Cantidad de registros filtrados en userBici22: 33
Cantidad de registros filtrados en userBici23: 95
Cantidad de registros filtrados en userBici24: 50


<strong>Eliminamos determinados registros de la columna edad_usuario <strong>

In [561]:
#Esta función elimina los registros que no cumplen con la condición de edad
def eliminar_registros_edad(df_list):
    for df in df_list:
        df.drop(df[(df['edad_usuario'] < 18) | (df['edad_usuario'] > 85)].index, inplace=True)
    return
listDfUserBici = [userBici15,userBici16, userBici17, userBici18, userBici19, userBici20, userBici21, userBici22, userBici23, userBici24]
eliminar_registros_edad(listDfUserBici)

In [562]:
userBici19['edad_usuario'].unique()

array([36, 43, 47, 45, 37, 23, 28, 31, 80, 41, 65, 25, 29, 30, 39, 56, 35,
       38, 32, 34, 48, 24, 58, 52, 55, 33, 46, 22, 27, 51, 42, 44, 54, 20,
       26, 66, 50, 21, 67, 79, 53, 40, 73, 59, 19, 57, 64, 60, 49, 62, 74,
       61, 85, 82, 63, 72, 76, 70, 69, 18, 68, 83, 71, 75, 78, 81, 77, 84],
      dtype=int64)

<strong>Definimos si en los registros tiene datos NaN (de manera de ejemplo solo se muestra de userBici15), pero se revisó en todos los df</strong>

In [563]:
#Esta función lee por cada columna si el registro es NaN 
def registros_nulos(df):
    filas_nulas = df[df.isnull().any(axis=1)]
    return filas_nulas
df_registros = registros_nulos(userBici15)
print(df_registros)

       id_usuario genero_usuario  edad_usuario fecha_alta    hora_alta
1954       132123            NaN            32   06-04-15   2:14:40 PM
7311       131922            NaN            38   08-04-15   3:08:44 AM
7510       147241            NaN            60   10-04-15  12:06:27 PM
7511       147254            NaN            60   10-04-15  12:13:48 PM
7785       149570            NaN            33   13-04-15   1:54:06 PM
...           ...            ...           ...        ...          ...
39989      222353            NaN            62   21-12-15  11:32:20 AM
39990      222354            NaN            57   21-12-15  11:36:47 AM
40315      223630            NaN            21   29-12-15   6:05:42 PM
40316      223631            NaN            27   29-12-15   6:08:35 PM
40318      223633            NaN            24   29-12-15   6:12:32 PM

[130 rows x 5 columns]


<strong>userbici15 se encontró </strong>    <span style='color: #FF6666;'>130 datos NaN</span> <strong> en la columna genero_usuario</strong>
<br><strong>userbici19 se encontró </strong><span style='color: #FF6666;'>64 datos NaN</span> <strong> en las columnas genero_usuario y edad_usuario<strong>


<strong>En este caso utilicé las funciones mediana y moda para agregar datos faltantes al ser poco datos no es necesario un modelo </strong>

In [564]:
# Para 'genero_usuario' usamos la moda = al valor más frecuente
moda_genero_15 = userBici15['genero_usuario'].mode()[0]
moda_genero_19 = userBici19['genero_usuario'].mode()[0]
#aplicamos la mediana en los df 
userBici15['genero_usuario'] = userBici15['genero_usuario'].fillna(moda_genero_15)
userBici19['genero_usuario'] = userBici19['genero_usuario'].fillna(moda_genero_19)

#Para 'edad_usuario' usar la mediana = el valor medio 
mediana_edad_19 = userBici19['edad_usuario'].median()
#aplicamos la mediana en los df 
userBici19['edad_usuario'] = userBici19['edad_usuario'].fillna(mediana_edad_19)

In [567]:
#Reviso lo registros únicos de la columna 'Customer.Has.Dni..Yes...No.' 
userBici23['Customer.Has.Dni..Yes...No.'].unique()

array(['Yes', 'No'], dtype=object)

<strong>En userBici20, userBici21, UsuerBici22, userBici23, userBici24 tiene una columna llamada 'Customer.Has.Dni..Yes...No.' que me retorna YES/NO sobre si tiene dni; En mi analisís y por las consginas determidas lo elimino porque me parece irrelevante <strong>

In [568]:
userBici20 = userBici20.drop(columns=['Customer.Has.Dni..Yes...No.'])
userBici21 = userBici21.drop(columns=['Customer.Has.Dni..Yes...No.'])
userBici22 = userBici22.drop(columns=['Customer.Has.Dni..Yes...No.'])
userBici23 = userBici23.drop(columns=['Customer.Has.Dni..Yes...No.'])
userBici24 = userBici24.drop(columns=['Customer.Has.Dni..Yes...No.'])
#verificamos en un df 
userBici24.columns

Index(['ID_usuario', 'genero_usuario', 'edad_usuario', 'fecha_alta',
       'hora_alta'],
      dtype='object')

<strong>En los df userbici15, userbici16, userbici17, userbici18 los valores de la columna genero_usuario son:<span style='color: #FF6666;'>['M', 'F','O']</span> </strong>
<br><strong>y en los demás están más detallados: <span style='color: #FF6666;'>['OTHER', 'MALE', 'FEMALE']</span>; Dejaré todos del mismo formato<strong>


In [569]:
#Defino una función para reemplazar los valores de la columna 'genero_usuario'
def reemplazar_genero(dataframes):
    for df in dataframes:
        df['genero_usuario'] = df['genero_usuario'].replace({'M': 'MALE', 'F': 'FEMALE', 'O': 'OTHER'})
    return 
reemplazar_genero([userBici15,userBici16,userBici17,userBici18])

<strong>Revisamos los cambios</strong>

In [570]:
userBici15['genero_usuario'].unique()

array(['MALE', 'FEMALE'], dtype=object)

In [571]:
userBici16['genero_usuario'].unique()

array(['MALE', 'FEMALE'], dtype=object)

In [572]:
userBici17['genero_usuario'].unique()

array(['MALE', 'FEMALE'], dtype=object)

In [573]:
userBici18['genero_usuario'].unique()

array(['FEMALE', 'MALE', 'OTHER'], dtype=object)

<strong>Le cambiamos el nombre de la columna 'ID_usuario' para que puedan coincidir <strong>

In [574]:
def replace_head(df):
    df = df.rename(columns = {'ID_usuario' : 'id_usuario'},inplace=True)
    return
replace_head (userBici19)
replace_head (userBici20)
replace_head (userBici21)
replace_head (userBici22)
replace_head (userBici23)
replace_head (userBici24)

In [575]:
userBici15['hora_alta']

0        5:32:55 PM
1        5:33:12 PM
2        5:35:14 PM
3        5:35:43 PM
4        5:35:43 PM
            ...    
40392    5:14:39 PM
40393    5:15:25 PM
40394    5:17:50 PM
40395    5:18:55 PM
40396    6:52:35 PM
Name: hora_alta, Length: 40360, dtype: object

<strong> Nos encontramos que la columna hora_alta en los df: userBici15, userBici16, userBici17, userBici18 no coinciden con los registros de de los demas df <strong>
<br><strong>Estandarizamos la columna hora_alta para que todos los registros coincidan<strong>

In [576]:
def estandarizar_hora(hora):
    try:
        return pd.to_datetime(hora, format='%I:%M:%S %p', errors='coerce').strftime('%H:%M:%S')
    except:
        try:
            return pd.to_datetime(hora, format='%H:%M:%S', errors='coerce').strftime('%H:%M:%S')
        except:
            return None

for df in [userBici15, userBici16, userBici17, userBici18]:
    df['hora_alta'] = df['hora_alta'].apply(estandarizar_hora)

#Comprobamos que la función haya funcionado correctamente
print(f" Userbici15 \n {userBici15['hora_alta'].head(1)}")
print(f" Userbici16 \n {userBici16['hora_alta'].head(1)}")
print(f" Userbici17 \n {userBici17['hora_alta'].head(1)}")
print(f" Userbici18 \n {userBici18['hora_alta'].head(1)}") 

 Userbici15 
 0    17:32:55
Name: hora_alta, dtype: object
 Userbici16 
 0    03:45:19
Name: hora_alta, dtype: object
 Userbici17 
 0    01:32:11
Name: hora_alta, dtype: object
 Userbici18 
 0    05:57:40
Name: hora_alta, dtype: object


<strong>Estandarizamos la columna fecha_alta para que todos los registros coincidan <strong>

In [577]:
def unificar_formato_fecha(dataframes):
    for df in dataframes:
        df['fecha_alta'] = pd.to_datetime(df['fecha_alta'], errors='coerce', format='%d-%m-%y').fillna(
            pd.to_datetime(df['fecha_alta'], errors='coerce', format='%Y-%m-%d')
        )
        df['fecha_alta'] = df['fecha_alta'].dt.strftime('%Y-%m-%d')
    return dataframes

# Aplicar la función
listDfUserBici_f = [userBici15, userBici16, userBici17, userBici18]
dataframes = unificar_formato_fecha(listDfUserBici_f)
print(f" Userbici15 \n {userBici15['fecha_alta'].head(1)}")
print(f" Userbici16 \n {userBici16['fecha_alta'].head(1)}")
print(f" Userbici17 \n {userBici17['fecha_alta'].head(1)}")
print(f" Userbici18 \n {userBici18['fecha_alta'].head(1)}")

 Userbici15 
 0    2015-02-28
Name: fecha_alta, dtype: object
 Userbici16 
 0    2016-01-01
Name: fecha_alta, dtype: object
 Userbici17 
 0    2017-01-01
Name: fecha_alta, dtype: object
 Userbici18 
 0    2018-10-03
Name: fecha_alta, dtype: object


<strong>Comparo con el userBici19 <strong>

In [578]:
print(f" Userbici19 \n {userBici19['fecha_alta'].head(1)}")

 Userbici19 
 0    2019-12-31
Name: fecha_alta, dtype: object


<strong>Verificamos si el nomnbre de las columnas y el tipo de las columnas son iguales<strong>

In [579]:
#Lista de DataFrames a verificar
listDfBici = [userBici15, userBici16, userBici17, userBici18, userBici19, userBici20, userBici21, userBici22, userBici23, userBici24]
#LOs ordenamos por id_usuario
for df in listDfBici:
    df.sort_values(by='id_usuario', ascending=True, inplace=True)
# Verifamos si todas las columnas son iguales en todos los DataFrames
columnas_iguales = all(df.columns.tolist() == listDfBici[0].columns.tolist() for df in listDfBici)

# Verificamos si todos los tipos de datos son iguales en todos los df
tipos_datos_iguales = all(df.dtypes.equals(listDfBici[0].dtypes) for df in listDfBici)

# Verificamos si todas las columnas y tipos de datos coinciden
if columnas_iguales and tipos_datos_iguales:
    print("Todos los DataFrames coinciden y se pueden unificar.")
else:
    print("Los DataFrames no coinciden y no se pueden unificar.")

Todos los DataFrames coinciden y se pueden unificar.


In [580]:
userBici = pd.concat(listDfBici, ignore_index=True)
#Revisamos la cantidad de registros en userBici
print(f"Tenemos {userBici.shape[0]} registros en userBici")

Tenemos 795728 registros en userBici


<strong>Verifico si tengo duplicados en base a la columna id_usuario<strong>

In [581]:
duplicados_ecoBici = userBici[userBici.duplicated(subset='id_usuario', keep=False)]
if not duplicados_ecoBici.empty:
    print(f"Se encontró {duplicados_ecoBici.shape[0]} registros duplicados")
else:
    print("No se encontraron registros duplicados en userBici.")

Se encontró 54980 registros duplicados


<strong>Realizamos un df_copia solo con los datos duplicados<strong>

In [582]:
def obtener_duplicado(df):
    duplicados = df[df.duplicated(subset='id_usuario', keep='first')]
    duplicado = duplicados.drop_duplicates(subset='id_usuario', keep='first')
    return duplicado

df_userBici_duplicado = obtener_duplicado(userBici)
df_userBici_duplicado = df_userBici_duplicado.sort_values(by='id_usuario', ascending=True)
print(f"Tenemos {df_userBici_duplicado.shape[0]} registros en df_userBici_duplicado")

Tenemos 27489 registros en df_userBici_duplicado


In [583]:
#Revisamos el valor máximo en la columna 'id_usuario'
valor_maximo = userBici['id_usuario'].max()
print(f"El valor más alto en la columna 'id_usuario' es: {valor_maximo}")

El valor más alto en la columna 'id_usuario' es: 1173048


In [584]:
usserBici = userBici.drop_duplicates(subset='id_usuario', inplace=True)
#Revisamos la cantidad de registros en userBici
print(f"Tenemos {userBici.shape[0]} registros en df_userBici")

Tenemos 768237 registros en df_userBici


<strong>Tenemos <span style='color: #FF6666;'> 768237 </span> datos limpios</strong>

In [585]:
print(f"El id_usuario más alto en userBici es: {userBici['id_usuario'].max()}")

El id_usuario más alto en userBici es: 1173048


<strong>Para no perder esos datos duplicados, en base al máximo id_usuario vamos a asignarle nuevos valores a partir de 1200000 para poder identificarlos </strong>

In [586]:
id_inicial = 1200000
df_userBici_duplicado['id_usuario'] = range(id_inicial, id_inicial + len(df_userBici_duplicado))

# Verificamos si se actualizaron los ids
df_userBici_duplicado['id_usuario'].tail(10)

319434    1227479
319438    1227480
319532    1227481
319537    1227482
319539    1227483
319902    1227484
319906    1227485
319917    1227486
319922    1227487
319957    1227488
Name: id_usuario, dtype: int64

<strong>Lo concatenamos con el df main</strong>

In [588]:
userBici = pd.concat([userBici, df_userBici_duplicado], ignore_index=False)
userBici = userBici.sort_values(by='id_usuario', ascending=True)

In [589]:
print(userBici['id_usuario'].dtype)
print(f"El valor más alto en la columna 'id_usuario' es: {userBici['id_usuario'].max()}")

int64
El valor más alto en la columna 'id_usuario' es: 1227488


<strong>Lo migramos a la base stg<strong>

In [590]:
conexion = sql3.connect('DataIQ_STG.db', timeout=10)

try:
    userBici.to_sql('usuarios_ecobici', conexion, if_exists='replace', index=False)
    conexion.commit()

except sql3.DatabaseError as e:
    print(f"Error en la base de datos: {e}")

finally:
    conexion.close()

<strong>Continuamos con el analisis de trips<strong>

In [135]:
trips24 = pd.read_csv("trips_2024.csv")
trips23 = pd.read_csv("trips_2023.csv")

<strong>Revisamos cuantos registros vienen en cada df</strong>

In [591]:
print(f"El archivo 2023 tiene {trips23.shape[0]} registos")
print(f"El archivo 2024 tiene {trips24.shape[0]} registros")

El archivo 2023 tiene 2622331 registos
El archivo 2024 tiene 1528727 registros


<strong>Reviso el tipo de dato de todas la columnas de los df<strong>

In [592]:
print(f"Trips 2023: \n{trips23.dtypes}")
print("-------------------")
print(f"Trips 2024: \n{trips24.dtypes}")

Trips 2023: 
Id_recorrido                   object
duracion_recorrido             object
fecha_origen_recorrido         object
id_estacion_origen             object
nombre_estacion_origen         object
direccion_estacion_origen      object
long_estacion_origen          float64
lat_estacion_origen           float64
fecha_destino_recorrido        object
id_estacion_destino            object
nombre_estacion_destino        object
direccion_estacion_destino     object
long_estacion_destino         float64
lat_estacion_destino          float64
id_usuario                     object
modelo_bicicleta               object
género                         object
dtype: object
-------------------
Trips 2024: 
Id_recorrido                   object
duracion_recorrido             object
fecha_origen_recorrido         object
id_estacion_origen             object
nombre_estacion_origen         object
direccion_estacion_origen      object
long_estacion_origen          float64
lat_estacion_origen         

<strong>Definimos si en los registros tiene datos NaN (de manera de ejemplo solo se muestra de trips23), pero se revisó en los dos df</strong>

In [597]:
#Llamamos a la funcion registros_nulos anteriormente definida
df_registros = registros_nulos(trips23)
print(df_registros)

              Id_recorrido duracion_recorrido fecha_origen_recorrido  \
114      18700657BAEcobici                964    2023-08-02 08:10:36   
190      18061078BAEcobici                552    2023-05-10 13:38:53   
229      17194819BAEcobici                261    2023-02-09 17:36:30   
727      16874429BAEcobici              3,542    2023-01-09 18:31:53   
1116     16973753BAEcobici                863    2023-01-19 09:01:24   
...                    ...                ...                    ...   
2621218  20050747BAEcobici              1,304    2023-12-16 16:12:52   
2621219  20050003BAEcobici              2,412    2023-12-16 13:19:49   
2621220  20049171BAEcobici                452    2023-12-16 10:14:50   
2621379  20154013BAEcobici                768    2023-12-28 13:45:51   
2621398  20158544BAEcobici                663    2023-12-28 19:40:09   

        id_estacion_origen            nombre_estacion_origen  \
114           220BAEcobici                097 - Avelino Diaz   
190    

<strong>trips23 se encontró </strong>    <span style='color: #FF6666;'>11968 datos NaN</span> <strong> en la columna género</strong>
<br><strong>trips24 se encontró </strong><span style='color: #FF6666;'>5465 datos NaN</span> <strong> en las columnas género <strong>

<strong>En este caso utilicé las funciones moda para agregar datos faltantes </strong>

In [599]:
# Para la columna 'género' usamos la moda = al valor más frecuente
moda_genero_23 = trips23['género'].mode()[0]
moda_genero_24 = trips24['género'].mode()[0]
#aplicamos la mediana en los df 
trips23['género'] = trips23['género'].fillna(moda_genero_23)
trips24['género'] = trips24['género'].fillna(moda_genero_24)

<strong>Verficamos si tienen registros iguales<strong>

In [600]:
trips23.equals(trips24)

False

<strong>Procedemos a concatenarlo</strong>

In [601]:
trips = pd.concat([trips23, trips24], ignore_index= True)
trips = trips.sort_values(by='Id_recorrido', ascending=True)
trips.head(10)

Id_recorrido duracion_recorrido fecha_origen_recorrido  \
1569107  16798820BAEcobici                392    2023-01-01 00:06:08   
1678989  16798822BAEcobici                250    2023-01-01 00:08:12   
1633091  16798823BAEcobici              1,059    2023-01-01 00:10:52   
782258   16798824BAEcobici              1,834    2023-01-01 00:11:14   
547929   16798825BAEcobici                766    2023-01-01 00:11:24   
185695   16798826BAEcobici                623    2023-01-01 00:13:14   
1545854  16798827BAEcobici              1,070    2023-01-01 00:15:09   
42259    16798828BAEcobici              1,194    2023-01-01 00:15:43   
1025713  16798829BAEcobici              3,142    2023-01-01 00:15:57   
136062   16798830BAEcobici              3,016    2023-01-01 00:16:45   

        id_estacion_origen          nombre_estacion_origen  \
1569107       449BAEcobici        352 - San Jose de Flores   
1678989       449BAEcobici        352 - San Jose de Flores   
1633091       363BAEcobici                354 - José Martí   
782258        461BAEcobici  106 - MOSCONI Y CONSTITUYENTES   
547929        124BAEcobici                 124 - UGARTECHE   
185695        307BAEcobici                  341 - COMUNA 4   
1545854       184BAEcobici             184 - Pedro Echague   
42259          13BAEcobici                      013 - ONCE   
1025713       152BAEcobici           152 - JULIETA LANTERI   
136062        152BAEcobici           152 - JULIETA LANTERI   

                            direccion_estacion_origen  long_estacion_origen  \
1569107                            Av. Rivadavia 6955            -58.463297   
1678989                            Av. Rivadavia 6955            -58.463297   
1633091                                Eva Perón 3321            -58.463114   
782258                            CONSTITUYENTES 4710            -58.493630   
547929                                 3050 Ugarteche            -58.411965   
185695                          AV PERITO MORENO 2881            -58.424773   
1545854     Pedro Echague 1878 & Combate De Los Pozos            -58.391342   
42259                  Av. Pueyrredon & Av. Rivadavia            -58.406000   
1025713  Vera Peñaloza, Rosario 599 & Lanteri Julieta            -58.359631   
136062   Vera Peñaloza, Rosario 599 & Lanteri Julieta            -58.359631   

         lat_estacion_origen fecha_destino_recorrido id_estacion_destino  \
1569107           -34.628836     2023-01-01 00:12:40        449BAEcobici   
1678989           -34.628836     2023-01-01 00:12:22        449BAEcobici   
1633091           -34.643900     2023-01-01 00:28:31        363BAEcobici   
782258            -34.580880     2023-01-01 00:41:48        418BAEcobici   
547929            -34.580538     2023-01-01 00:24:10        131BAEcobici   
185695            -34.649997     2023-01-01 00:23:37         41BAEcobici   
1545854           -34.630613     2023-01-01 00:32:59          3BAEcobici   
42259             -34.610090     2023-01-01 00:35:37        122BAEcobici   
1025713           -34.618164     2023-01-01 01:08:19         25BAEcobici   
136062            -34.618164     2023-01-01 01:07:01         25BAEcobici   

           nombre_estacion_destino  \
1569107   352 - San Jose de Flores   
1678989   352 - San Jose de Flores   
1633091           354 - José Martí   
782258         345 - Plaza Mafalda   
547929   131- HOSPITAL DE CLÍNICAS   
185695   041 - PARQUE PATRICIOS II   
1545854               003 - ADUANA   
42259             122 - COSTA RICA   
1025713         025 - Plaza Guemes   
136062          025 - Plaza Guemes   

                                direccion_estacion_destino  \
1569107                                 Av. Rivadavia 6955   
1678989                                 Av. Rivadavia 6955   
1633091                                     Eva Perón 3321   
782258                           Conde & Concepción Arenal   
547929                   Uriburu Jose E., Pres. & Paraguay   
185695                                  

<strong>Determinamos dejar en el id_usuario, id_recorrido, id_estacion_origen y id_estacion_destino solo los números<strong>

In [602]:
trips['Id_recorrido'] = trips['Id_recorrido'].str.extract('(\d+)')
trips['id_estacion_origen'] = trips['id_estacion_origen'].str.extract('(\d+)')
trips['id_estacion_destino'] = trips['id_estacion_destino'].str.extract('(\d+)')
trips['id_usuario'] = trips['id_usuario'].str.extract('(\d+)')
#Revisamos los cambios
trips.head()

Id_recorrido duracion_recorrido fecha_origen_recorrido  \
1569107     16798820                392    2023-01-01 00:06:08   
1678989     16798822                250    2023-01-01 00:08:12   
1633091     16798823              1,059    2023-01-01 00:10:52   
782258      16798824              1,834    2023-01-01 00:11:14   
547929      16798825                766    2023-01-01 00:11:24   

        id_estacion_origen          nombre_estacion_origen  \
1569107                449        352 - San Jose de Flores   
1678989                449        352 - San Jose de Flores   
1633091                363                354 - José Martí   
782258                 461  106 - MOSCONI Y CONSTITUYENTES   
547929                 124                 124 - UGARTECHE   

        direccion_estacion_origen  long_estacion_origen  lat_estacion_origen  \
1569107        Av. Rivadavia 6955            -58.463297           -34.628836   
1678989        Av. Rivadavia 6955            -58.463297           -34.628836   
1633091            Eva Perón 3321            -58.463114           -34.643900   
782258        CONSTITUYENTES 4710            -58.493630           -34.580880   
547929             3050 Ugarteche            -58.411965           -34.580538   

        fecha_destino_recorrido id_estacion_destino  \
1569107     2023-01-01 00:12:40                 449   
1678989     2023-01-01 00:12:22                 449   
1633091     2023-01-01 00:28:31                 363   
782258      2023-01-01 00:41:48                 418   
547929      2023-01-01 00:24:10                 131   

           nombre_estacion_destino         direccion_estacion_destino  \
1569107   352 - San Jose de Flores                 Av. Rivadavia 6955   
1678989   352 - San Jose de Flores                 Av. Rivadavia 6955   
1633091           354 - José Martí                     Eva Perón 3321   
782258         345 - Plaza Mafalda          Conde & Concepción Arenal   
547929   131- HOSPITAL DE CLÍNICAS  Uriburu Jose E., Pres. & Paraguay   

         long_estacion_destino  lat_estacion_destino id_usuario  \
1569107             -58.463297            -34.628836     703716   
1678989             -58.463297            -34.628836     695884   
1633091             -58.463114            -34.643900     671615   
782258              -58.444580            -34.580938     939857   
547929              -58.399016            -34.598404     721955   

        modelo_bicicleta género  
1569107              FIT   MALE  
1678989              FIT   MALE  
1633091              FIT   MALE  
782258               FIT   MALE  
547929            ICONIC   MALE

<strong>Verificamos si tenemos duplicados condicionandolo por id_recorrido<strong>

In [603]:
duplicados_trips = trips[trips.duplicated(subset='Id_recorrido', keep=False)]
if not duplicados_trips.empty:
    print(f"Se encontró {duplicados_trips.shape[0]} registros duplicados")
else:
    print("No se encontraron registros duplicados en trips.")

No se encontraron registros duplicados en trips.


In [604]:
trips.dtypes

Id_recorrido                   object
duracion_recorrido             object
fecha_origen_recorrido         object
id_estacion_origen             object
nombre_estacion_origen         object
direccion_estacion_origen      object
long_estacion_origen          float64
lat_estacion_origen           float64
fecha_destino_recorrido        object
id_estacion_destino            object
nombre_estacion_destino        object
direccion_estacion_destino     object
long_estacion_destino         float64
lat_estacion_destino          float64
id_usuario                     object
modelo_bicicleta               object
género                         object
dtype: object

<strong>Lo dejamos a las columnas que son referencia a un id en tipo int<strong>

In [607]:
trips['Id_recorrido'] = trips['Id_recorrido'].astype(int)
trips['id_estacion_origen'] = trips['id_estacion_origen'].astype(int)
trips['id_estacion_destino'] = trips['id_estacion_destino'].fillna(0).astype(int)
trips['id_usuario'] = trips['id_usuario'].astype(int)

<strong>Lo migramos a la base staging <strong>

In [609]:
conexion = sql3.connect('DataIQ_STG.db', timeout=10)

try:
    trips.to_sql('trips', conexion, if_exists='replace', index=False)
    conexion.commit()

except sql3.DatabaseError as e:
    print(f"Error en la base de datos: {e}")

finally:
    conexion.close()

<strong>Analizamos nuevas-estaciones-bicicletas-publicas y retornamos cuantos registros tiene</strong>

In [610]:
newStation = pd.read_csv("nuevas-estaciones-bicicletas-publicas.csv")
print(f"El archivo tiene {newStation.shape[0]} registros")

El archivo tiene 360 registros


In [617]:
newStation.head()

NÚMERO DE ESTACIÓN               NOMBRE  \
136                   1  FACULTAD DE DERECHO   
0                     2             RETIRO I   
1                     3               ADUANA   
2                     4           PLAZA ROMA   
3                     5         PLAZA ITALIA   

                               DIRECCIÓN       BARRIO     COMUNA  \
136      Av. Pres. Figueroa Alcorta 2120     RECOLETA   COMUNA 2   
0    Av. Dr. José María Ramos Mejía 1300       RETIRO   COMUNA 1   
1                               Moreno 4    MONSERRAT   COMUNA 1   
2                     Av. Corrientes 100  SAN NICOLAS   COMUNA 1   
3                     Av. Sarmiento 2601      PALERMO  COMUNA 14   

    EMPLAZAMIENTO                  LATITUD                LONGITUD  
136        VEREDA                 -345.841                -583.908  
0           PLAZA  -34.592.424.135.701.100  -5.837.470.989.599.140  
1          VEREDA                 -346.111                -583.682  
2          VEREDA   -3.460.299.498.843.600  58.368.832.321.694.600  
3          VEREDA                 -345.807                -584.204

In [611]:
newStation.dtypes

ID Comet               int64
NÚMERO de Estación     int64
NOMBRE                object
DIRECCIÓN             object
BARRIO                object
COMUNA                object
EMPLAZAMIENTO         object
LATITUD               object
LONGITUD              object
dtype: object

<strong>Definimos si en los registros tiene datos NaN </strong>

In [612]:
#Llamamos a la funcion registros_nulos anteriormente definida
df_registros = registros_nulos(newStation)
print(df_registros)

Empty DataFrame
Columns: [ID Comet, NÚMERO de Estación, NOMBRE, DIRECCIÓN, BARRIO, COMUNA, EMPLAZAMIENTO, LATITUD, LONGITUD]
Index: []


In [613]:
newStation['ID Comet'].unique()

array([  2,   3,   4,   5,   6,   7,   8,   9,  12,  13,  14,  17,  21,
        22,  23,  24,  25,  26,  27,  29,  30,  32,  33,  35,  36,  38,
        41,  43,  44,  45,  46,  48,  49,  50,  51,  54,  56,  57,  58,
        59,  60,  61,  63,  64,  65,  66,  68,  69,  70,  71,  73,  74,
        75,  76,  77,  79,  80,  82,  83,  85,  86,  87,  89,  91,  92,
        93,  94,  95,  96,  98,  99, 101, 102, 104, 107, 111, 112, 114,
       116, 117, 118, 120, 121, 122, 124, 126, 128, 130, 131, 132, 134,
       135, 137, 138, 144, 146, 149, 150, 151, 152, 153, 155, 156, 158,
       161, 162, 163, 164, 165, 166, 167, 168, 169, 171, 172, 174, 175,
       176, 177, 179, 181, 182, 183, 184, 186, 187, 188, 189, 190, 191,
       193, 194, 196, 197, 199, 200, 202, 203, 204, 205, 206, 207, 208,
       210, 212, 213, 215, 216, 219, 220, 222, 223, 227, 228, 229, 230,
       231, 232, 234, 235, 236, 237, 239, 241, 242, 245, 247, 248, 251,
       252, 253, 254, 255, 257, 258, 259, 260, 261, 262, 263, 26

In [614]:
newStation['NÚMERO de Estación'].unique()

array([  2,   3,   4,   5,   6,   7,   8,   9,  12,  13,  14,  17,  21,
        22,  23,  24,  25,  26,  27,  29,  30,  32,  33,  35,  36,  38,
        41, 334,  44,  45,  46,  34,  49,  50,  51,  54,  56,  57,  58,
        18,  60,  98,  63,  64,  65,  66, 105,  69,  70,  71,  73,  74,
        75,  76, 109,  79,  80,  82,  83,  85,  86,  87,  89,  91,  92,
        93,  94,  95,  96,  37,  99, 101, 102, 104, 107, 111, 112, 114,
       116, 117, 118, 120, 121, 122, 124, 126, 128, 130, 131, 132, 134,
       135, 137, 138, 144, 146, 149, 150, 151, 152, 153, 155, 156, 158,
       161, 162, 163, 164, 165, 166, 275, 168, 169, 171, 172, 174, 147,
        67, 177, 179, 181, 182, 183, 184, 186, 261, 373, 189, 190, 191,
       193, 194, 196, 197, 199, 200,   1, 108, 167, 125, 159, 123, 145,
       335, 129, 136, 113,  59, 140,  97, 160, 178, 188,  10, 243, 259,
       384, 252, 344, 281, 254, 302, 391, 348, 103, 248, 282, 346, 388,
       331, 383, 276, 374, 264, 336, 290, 371, 272, 349, 270, 28

<strong>Eliminamos la columna'ID Comet' y modificamos un detalle en la columna estación<strong>

In [615]:
newStation = newStation.drop(columns=['ID Comet'])
newStation = newStation.rename(columns={'NÚMERO de Estación': 'NÚMERO DE ESTACIÓN'})
newStation = newStation.sort_values(by='NÚMERO DE ESTACIÓN', ascending=True)
newStation.head()

NÚMERO DE ESTACIÓN               NOMBRE  \
136                   1  FACULTAD DE DERECHO   
0                     2             RETIRO I   
1                     3               ADUANA   
2                     4           PLAZA ROMA   
3                     5         PLAZA ITALIA   

                               DIRECCIÓN       BARRIO     COMUNA  \
136      Av. Pres. Figueroa Alcorta 2120     RECOLETA   COMUNA 2   
0    Av. Dr. José María Ramos Mejía 1300       RETIRO   COMUNA 1   
1                               Moreno 4    MONSERRAT   COMUNA 1   
2                     Av. Corrientes 100  SAN NICOLAS   COMUNA 1   
3                     Av. Sarmiento 2601      PALERMO  COMUNA 14   

    EMPLAZAMIENTO                  LATITUD                LONGITUD  
136        VEREDA                 -345.841                -583.908  
0           PLAZA  -34.592.424.135.701.100  -5.837.470.989.599.140  
1          VEREDA                 -346.111                -583.682  
2          VEREDA   -3.460.299.498.843.600  58.368.832.321.694.600  
3          VEREDA                 -345.807                -584.204

In [616]:
duplicados_newStation = newStation[newStation.duplicated(subset='NÚMERO DE ESTACIÓN', keep=False)]
if not duplicados_newStation.empty:
    print(f"Se encontró {duplicados_newStation.shape[0]} registros duplicados")
else:
    print("No se encontraron registros duplicados en newStation.")

No se encontraron registros duplicados en newStation.


<strong>Lo migramos a la base staging<strong>

In [620]:
conexion = sql3.connect('DataIQ_STG.db', timeout=10)

try:
    newStation.to_sql('nuevas_estaciones_bicicletas', conexion, if_exists='replace', index=False)
    conexion.commit()

except sql3.DatabaseError as e:
    print(f"Error en la base de datos: {e}")

finally:
    conexion.close()

<strong>Lo siguiente que se realizará es desde la base Data_STG poblaremos los datos a la base Data_DWH<strong>